Documentation/designs/status-subresource.md

```diff
@@ -0,0 +1,436 @@
+
+```golang
+type PrometheusStatus struct {
+	// Represents whether any actions on the underlying managed objects are
+	// being performed. Only delete actions will be performed.
+	Paused bool `json:"paused"`
+	// Total number of non-terminated pods targeted by this Prometheus deployment
+	// (their labels match the selector).
+	Replicas int32 `json:"replicas"`
+	// Total number of non-terminated pods targeted by this Prometheus deployment
+	// that have the desired version spec.
+	UpdatedReplicas int32 `json:"updatedReplicas"`
+	// Total number of available pods (ready for at least minReadySeconds)
+	// targeted by this Prometheus deployment.
+	AvailableReplicas int32 `json:"availableReplicas"`
+	// Total number of unavailable pods targeted by this Prometheus deployment.
+	UnavailableReplicas int32 `json:"unavailableReplicas"`
+	// The current state of the Prometheus service.
+	// +patchMergeKey=type
+	// +patchMergeStrategy=merge
+	// +optional
+	Conditions []PrometheusStatusCondition `json:"conditions,omitempty"`
+	// The list has one entry per shard. Each entry provides a summary of the shard status.
+	// +patchMergeKey=shardID
+	// +patchMergeStrategy=merge
+	// +optional
+	ShardStatuses []ShardStatus `json:"shardStatuses,omitempty"`
+}
+
+// PrometheusStatusCondition represents the state of the resources associated with the Prometheus resource.
+// +k8s:deepcopy-gen=true
+type PrometheusStatusCondition struct {
+	// Type of the condition being reported.
+	// +required
+	Type PrometheusConditionType `json:"type"`
+	// status of the condition.
+	// +required
+	Status PrometheusConditionStatus `json:"status"`
+	// lastTransitionTime is the time of the last update to the current status property.
+	// +required
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+	// Reason for the condition's last transition.
+	// +optional
+	Reason *string `json:"reason,omitempty"`
+	// Human-readable message indicating details for the condition's last transition.
+	// +optional
+	Message *string `json:"reason,omitempty"`
+}
+
+type PrometheusConditionType string
+
+const (
+	// Available indicates that enough Prometheus pods are ready to provide the service.
+	PrometheusAvailable PrometheusConditionType = "Available"
+	// Degraded indicates that some Prometheus pods don't run as expected.
+	PrometheusDegraded PrometheusConditionType = "Degraded"
+	// Reconciled indicates that the operator has reconciled the state of the underlying resources with the Prometheus object spec.
+	PrometheusReconciled PrometheusConditionType = "Reconciled"
+)
+
+type PrometheusConditionStatus string
+
+const (
+	PrometheusConditionTrue    PrometheusConditionStatus = "True"
+	PrometheusConditionFalse   PrometheusConditionStatus = "False"
+	PrometheusConditionUnknown PrometheusConditionStatus = "Unknown"
+)
+
+type ShardStatus struct {
+	// Identifier of the shard.
+	// +required
+	ShardID string `json:"shardID"`
```

Comment:
Is there some association from a shard to a pod? For example, if there's something wrong with a particular shard, how would the user find the kubernetes pod for this shard?

Reply:
I expected the information to be present in the Message field of the top-level condition type (e.g. `"shard 0: prometheus-monitoring-0 not ready"`).

Reply:
Ok, that makes sense 👍